File: data_analysis_tools/data_cleaning.js
// data_cleaning.js

const { ipcRenderer } = require('electron');

let file_path = '';

export async function data_cleaning() {

    // 獲取 DOM 元素
    const contentDiv = document.getElementById('content');

    contentDiv.innerHTML = `

        <h1>数据清洗</h1>
        
        <div class="import">
            <div>
                <label>源数据文件</label>
                <input id="file_path" type="text">
            </div>

            <div>
                <label for="sheetDropdown">选择工作表</label>
                <select id="sheetDropdown" name="sheetDropdown"></select>
            </div>

            <div>
                <label for="columnDropdown">选择要进行清洗的列</label>
                <select id="columnDropdown" name="columnDropdown"></select>
            </div>

            <div>
                <label for="cleaning_modeDropdown">选择清理模式</label>
                <select id="cleaning_modeDropdown" name="cleaning_modeDropdown">
                    <option value="remove_duplicates">删除重复行</option>
                    <option value="fill_missing_zero">填充缺失值为：0</option>
                    <option value="fill_missing_blank">填充缺失值为：<空白></option>
                    <option value="fill_missing_previous">填充缺失值为：重复上一行</option>
                    <option value="standardize_text">标准化文本（去除首尾空格及转换为小写英文字母）</option>
                    <option value="convert_data_str">将数据类型转换为字符型</option>
                    <option value="convert_data_int">将数据类型转换为整数型</option>
                    <option value="convert_data_float">将数据类型转换为浮点数型</option>
                    <option value="convert_data_date">将数据类型转换为时间日期类型</option>
                </select>
            </div>

            <div>
                <button id="selectButton">选择文件</button>
                <button id="cleanButton">数据清洗</button>
                <button id="exportButton">导出SQLite数据库</button>
            </div>
        </div>

        <div class="export">
            <div>
                <label>操作反馈：</label>
            </div>
            <div>
                <textarea id="result_output" rows="18"></textarea>
            </div>
        </div>
    `;

    var file_path_input = document.getElementById('file_path');
    file_path_input.classList.add('readonly');
    file_path_input.readOnly = true;

    const sheetDropdown = document.getElementById('sheetDropdown');
    const columnDropdown = document.getElementById('columnDropdown');
    const cleaning_modeDropdown = document.getElementById('cleaning_modeDropdown');

    // 导入按钮js代码
    document.getElementById('selectButton').addEventListener('click', async () => {
        const fileFilters = [{ name: 'Excel Files', extensions: ['xlsx', 'xls'] }];
        file_path = await ipcRenderer.invoke('dialog:openFile', fileFilters);
        if (!file_path) { return }
 
        ipcRenderer.send('asynchronous-message', { command: 'data_cleaning_select', data: {'file_path': file_path }});
    });

    // 当用户选择工作表时，获取该工作表的列索引
    document.getElementById('sheetDropdown').addEventListener('change', async (event) => {
        const selected_sheet = event.target.value;

        ipcRenderer.send('asynchronous-message', { command: 'data_cleaning_index', data: {'file_path': file_path, 'sheet_name': selected_sheet }});
    });

    // 工作表选择的更改事件
    const triggerChange = (element) => {
        var event = new Event('change');
        element.dispatchEvent(event);
    };

    document.getElementById('cleanButton').addEventListener('click', async () => {

        file_path = file_path_input.value;
        if (!file_path) {
            alert('请先导入文件！');
            return;
        }
        
        const data = {
            file_path: file_path,
            sheet_name: sheetDropdown.value,
            column_name: columnDropdown.value,
            cleaning_mode: cleaning_modeDropdown.value
        };

        ipcRenderer.send('asynchronous-message', { command: 'data_cleaning_clean', data: data });
    });

    document.getElementById('exportButton').addEventListener('click', async () => {

        file_path = file_path_input.value;
        if (!file_path) {
            alert('请先导入文件！');
            return;
        }

        const data = {
            file_path: file_path,
            sheet_name: sheetDropdown.value,
        };

        ipcRenderer.send('asynchronous-message', { command: 'data_cleaning_export', data: data});
    });

    ipcRenderer.removeAllListeners('asynchronous-reply');

    ipcRenderer.on('asynchronous-reply', (event, result) => {

        if (result[0] === 'data_cleaning_select') {

            file_path_input.value = file_path;

            sheetDropdown.innerHTML = '';

            result[1].forEach(item => {
                const option = document.createElement('option');
                option.value = item;
                option.text = item;
                sheetDropdown.appendChild(option);
            });

            triggerChange(sheetDropdown);       // 触发 change 事件，自动加载列数据

            document.getElementById(`result_output`).value += '选择文件成功！\n';

        } else if (result[0] === 'data_cleaning_index') {
    
            columnDropdown.innerHTML = '';
    
            result[1].forEach(item => {
                const columnOption = document.createElement('option');
                columnOption.value = item;
                columnOption.text = item;
                columnDropdown.appendChild(columnOption);
            });

        } else if (result[0] === 'data_cleaning_clean') {

            document.getElementById(`result_output`).value += result[1];

        } else if (result[0] === 'data_cleaning_export') {

            document.getElementById(`result_output`).value += result[1];

        };
    });
}

File: data_analysis_tools/data_cleaning.py
# data_cleaning.py

# 数据清洗：对获取的数据进行清洗，确保其格式统一、无重复项。

import os
import shutil
import pandas as pd
import sqlite3

def data_cleaning_select(request):

    file_path = request.get("data", {}).get("file_path", "")

    sheet_file = pd.ExcelFile(file_path)                    # 使用pandas讀取Excel文件
    sheetnames = sheet_file.sheet_names                     # 獲取所有工作表名稱

    return ['data_cleaning_select', sheetnames]


def data_cleaning_index(request):

    file_path = request.get("data", {}).get("file_path", "")
    sheet_name = request.get("data", {}).get("sheet_name", "")

    file_extension = os.path.splitext(file_path)[1].lower()

    if file_extension == '.xlsx':
        df = pd.read_excel(file_path, sheet_name=sheet_name, engine='openpyxl')
    elif file_extension == '.xls':
        df = pd.read_excel(file_path, sheet_name=sheet_name, engine='xlrd')

    columns = df.columns.tolist()                           # 获取工作表的列名

    return ['data_cleaning_index', columns]


def data_cleaning_clean(request):
    
    file_path = request.get("data", {}).get("file_path", "")
    sheet_name = request.get("data", {}).get("sheet_name", "")
    column_name = request.get("data", {}).get("column_name", "")
    cleaning_mode = request.get("data", {}).get("cleaning_mode", "")

    folder_path = os.path.dirname(file_path)
    target_path = os.path.join(folder_path, 'data_cleaning.xlsx')

    if os.path.exists(target_path):

        df = pd.read_excel(target_path, engine='openpyxl')

    else:

        file_extension = os.path.splitext(file_path)[1].lower()

        if file_extension == '.xlsx':
            df = pd.read_excel(file_path, sheet_name=sheet_name, engine='openpyxl')
        elif file_extension == '.xls':
            df = pd.read_excel(file_path, sheet_name=sheet_name, engine='xlrd')

        df.to_excel(target_path, index=False)

    # 数据清理
    if cleaning_mode == 'remove_duplicates':
        df = df.drop_duplicates(subset=[column_name])               # 删除指定列的重复行

    elif cleaning_mode == 'fill_missing_zero':
        df[column_name] = df[column_name].fillna(0)                 # 填充缺失值为 0

    elif cleaning_mode == 'fill_missing_blank':
        df[column_name] = df[column_name].fillna('<空白>')           # 填充缺失值为 '<空白>'

    elif cleaning_mode == 'fill_missing_previous':
        df[column_name] = df[column_name].ffill()                   # 充缺失值为 重复上一行

    elif cleaning_mode == 'standardize_text':
        df[column_name] = df[column_name].str.strip().str.lower()   # 标准化文本（去除首尾空格及转换为小写英文字母）

    elif cleaning_mode == 'convert_data_str':
        df[column_name] = df[column_name].astype(str)               # 将数据类型转换为字符型

    elif cleaning_mode == 'convert_data_int':
        df[column_name] = df[column_name].astype(str).str.replace(',', '')
        df[column_name] = pd.to_numeric(df[column_name], errors='coerce').fillna(0).astype(int) # 转换为整数，支持缺失值

    elif cleaning_mode == 'convert_data_float':
        df[column_name] = df[column_name].astype(str).str.replace(',', '')
        df[column_name] = pd.to_numeric(df[column_name], errors='coerce').astype(float)         # 转换为浮点数，支持缺失值

    elif cleaning_mode == 'convert_data_date':
        df[column_name] = pd.to_datetime(df[column_name], errors='coerce')                      # 将数据类型转换为时间日期类型

    elif cleaning_mode == 'drop_columns':
        df = df.drop(columns=[column_name])                         # 删除指定列

    backup_path = os.path.join(folder_path, 'data_cleaning_1.xlsx')

    counter = 2
    while os.path.exists(backup_path):
        new_filename = f'data_cleaning_{counter}.xlsx'
        backup_path = os.path.join(folder_path, new_filename)
        counter += 1

    shutil.copy(target_path, backup_path)

    df.to_excel(target_path, index=False)

    result_text = f'\n数据清洗成功，清洗前数据备份文件路径：{backup_path}\n'

    return ['data_cleaning_clean', [result_text]]


def data_cleaning_export(request):

    file_path = request.get("data", {}).get("file_path", "")
    sheet_name = request.get("data", {}).get("sheet_name", "")

    folder_path = os.path.dirname(file_path)
    course_path = os.path.join(folder_path, 'data_cleaning.xlsx')
    target_path = os.path.join(folder_path, 'sqlite_database.db')

    if os.path.exists(course_path):

        df = pd.read_excel(course_path, engine='openpyxl')

    else:

        file_extension = os.path.splitext(file_path)[1].lower()

        if file_extension == '.xlsx':
            df = pd.read_excel(file_path, sheet_name=sheet_name, engine='openpyxl')
        elif file_extension == '.xls':
            df = pd.read_excel(file_path, sheet_name=sheet_name, engine='xlrd')
    
    # 將 DataFrame 輸出到 SQLite 資料庫
    try:

        conn = sqlite3.connect(target_path)                             # 建立或開啟資料庫檔案
        table_name = sheet_name                                         # 你可以自訂資料表名稱
        df.to_sql(table_name, conn, if_exists='replace', index=False)   # 寫入資料表
        conn.close()
        result_text = f'\n导出SQLite数据库成功，数据库文件路径：{target_path}\n'

    except Exception as e:

        result_text = f'\n导出失败，错误信息：{str(e)}\n'

    return ['data_cleaning_export', [result_text]]